ssl/t1_lib.c

```diff
@@ -3302,7 +3302,7 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
                 if ((lu = tls1_get_legacy_sigalg(s, -1)) == NULL) {
                     if (!fatalerrs)
                         return 1;
-                    SSLfatal(s, SSL_AD_INTERNAL_ERROR,
+                    SSLfatal(s, SSL_AD_HANDSHAKE_FAILURE,
```

Comment:
isn't the default SHA-1, and if SECLEVEL is set sufficiently high, it will fail?

Reply:
Yes.
Clients that want to work in a modern (web) environment need to not rely on the server using the defaults.
I'm not seeing an argument that this point in the code is anything other than an internal error, though.

Reply:
my point is that this happens when a legitimate client (that supports just RSA ciphersuites, doesn't advertise supported sig_algs, advertises TLS 1.2 as the highest supported version) connects to a legitimate server (that has support for SHA-1 disabled)

now, either in such situation we shouldn't get internal_error, or this code shouldn't get executed in the first place

do you mean the latter?

Reply:
Well, speaking formally, "insufficient_security" if we have fetched the defaults but rejected them because of SECLEVEL will fit better. But I tend not to distinguish it here.

Reply:
That's no longer a legitimate client in this day and age.
Effective participation in the modern TLS ecosystem requires sending supported sig_algs.
However, this is not the code that will cause the handshake failure in that scenario; that would be a few lines later when we actually check against the default values.  This is just doing a database lookup for what the default value should be, and that lookup "should" never fail.

Reply:
@kaduk If you look at tls1_get_legacy_sigalg(), you can see that there is a call to tls12_sigalg_allowed() which in turn calls the security level check. So this is definitely not just a database lookup and this change is IMHO correct.

Reply:
Okay, thanks for nudging me to look more closely at `tls1_get_legacy_sigalg()`, I do see the security-level checks at the end of the function.  That said, there are still plenty of codepaths that can return NULL for reasons other than the security check, so it seems to me that there are codepaths where either alert is more correct, and thus neither the current code nor the current state of the PR is truly "correct".

Factoring out the `tls12_sigalg_allowed()` call into `tls_choose_sigalg()` itself is not a trivial fix, since then it would duplicate work with `tls12_shared_sigalgs()`.  Pushing the SSLfatal() calls up into `tls1_get_legacy_sigalg()` would be more mechanical but doesn't seem terribly pretty either.

Reply:
IMO giving handshake failure in some cases where an unlikely internal error event happened is harmless. So I am OK with the PR as is.